File: LinkedList.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std ;


class Node()
{
    int data ;
    Node * next ;
    Node(int val)
    {
        data = val;
        next = NULL ;
    }
};
void insert(Node * &head , int val)
{
    Node * newNode = new Node(val);
    if(!head)
    {
        head = newNode ;
        return ;
    }
    Node * temp = head ;
    while(temp->next != NULL)
    {
        temp=temp->next ;
    }
    temp->next = newNode ;
}
void deleteNodeByValue(Node * &head , int val)
{
    Node * temp = head ;
    while(temp->next->data != val && temp!=NULL)
    {
        if(temp==head)
        {
            head=temp->next ;
            delete(temp);
            return ;
        }
        temp=temp->next ;
    }
    Node * n = temp->next ;
    temp->next = temp->next->next ;
    delete(n);
}
void deleteNodeByPosition(Node *&head, int n)
{
    Node *temp = head;
    if (n == 1)
    {
        head = temp->next;
        delete (temp);
        return;
    }
    for (int i = 1; i < n - 1; i++)
    {
        temp = temp->next;
    }
    Node *temp2 = temp->next;
    temp->next = temp2->next;
    delete (temp2);
}
void update(Node * &head , int pval , int nval )
{
    Node * temp = head ;
    while(temp->data != pval && temp!=NULL)
    {
        temp=temp->next ;
    }
    temp->data = nval ;
}
Node *merge(Node *a, Node *b)
{
    // base case
    if(a == NULL)
        return b;
    if(b == NULL)
        return a;
 
    // recursive case
    // take a head pointer
    Node *c;
 
    if(a->data < b->data)
    {
        c = a;
        c->next = merge(a->next, b);
    }
    else
    {
        c = b;
        c->next = merge(a, b->next);
    }
 
    return c;
}
 
Node *mid_point(Node *head)
{
    if(head == NULL || head->next == NULL)
        return head;

    Node *fast = head;
    Node *slow = head;
 
    while(fast != NULL && fast->next != NULL)
    {
        fast = fast->next;
 
        if(fast->next == NULL)
            break;
 
        fast = fast->next;
        slow = slow->next;
    }
 
    return slow;
}
 
Node* merge_sort(Node *head)
{
    // base case
    if(head == NULL || head->next == NULL)
        return head;
 
    // recursive case
    // Step 1: divide the linked list into
    // two equal linked lists
    Node *mid = mid_point(head);
    Node *a = head;
    Node *b = mid->next;
 
    mid->next = NULL;
 
    // Step 2: recursively sort the smaller
    // linked lists
    a = merge_sort(a);
    b = merge_sort(b);
 
    // Step 3: merge the sorted linked lists
    Node *c = merge(a, b);
 
    return c;
}
void print(Node * head)
{
    while(head!=NULL)
    {
        cout<<head->data<<endl;
        head=head->next ;
    }
}
int main()
{
    while(1){
    cout<<"Linked List"<<endl<<"Select Option"<<endl;
    cout<<"1. Add Element"<<endl<<"2. Delete Element"<<endl<<"3. Update List "<<endl<<"4. Sort the List"<<endl<<"5. Print List"<<endl<<"6. Exit";
    int choice = 0 ;
    cin>>choice;
    switch(choice)
    {
        case 1:
        cout<<"Enter the value to be inserted"<<endl;
        int val ; cin>>val ;
        insert(head,val);

        case 2:
        cout<<"Delete Element by"<<endl<<"1. position"<<endl<<"2. Value"<<endl;
        int ch = 0 ;
        cin>>ch;
        if(ch==2){
            cout<<"Enter the Value of the Element to be deleted"<<endl;
            int val ; cin>>val ;
            deleteNodeByValue(head , val);
        }
        else if(ch==1){
            cout<<"Enter the Positon of the Element to be deleted"<<endl;
            int pos ; cin>>pos ;
            deleteNodeByPosition(head , pos);
        }
        else{
            cout<<"invalid Choice"<<endl;
        }

        case 3:
        cout<<"Enter previous Value and New value to update the list ( Note : values should be separated by space ) "<<endl;
        int p_val , n_val ;
        cin>>p_val>>n_val;
        update(head , p_val , n_val);

        case 4:
        cout<<"Sort the List"<<endl;
        Node * n = merge_sort(head);
        head=n ;

        case 5:
        print(head);

        case 6:
        break ;
    }



}

File: Data Structures/LinkedList/linkedList.java
/*
 * Implementation of a linked list for Integers.
 *  This can be also be converted into a generic LinkedList by changing the
 *  type of the data to a generic type.
 */

// Importing Exception class for handling exceptions
import java.util.InputMismatchException;
//Importing Scanner class for user input
import java.util.Scanner;

public class linkedList {
    static class Node {
        //  Data of the node
        int data;
        // Pointer to the next node
        Node next;
        Node(int data){
            this.data = data;
            this.next = null;
        }
    }
    public static void main(String[] args) {
        // Create a blank linked list
        Node head = null;
        //  Creating a infinite loop to keep the program running
        while(true) {
            //  Printing the menu for the user
            System.out.println("--------- MENU ---------");
            System.out.println("Press 0 : Create a New LinkedList");
            System.out.println("Press 1 : Add a new Node at End");
            System.out.println("Press 2 : Add a new Node at First");
            System.out.println("Press 3 : Add a new Node at Position");
            System.out.println("Press 4 : Delete a Node");
            System.out.println("Press 5 : Update a Node");
            System.out.println("Press 6 : Sort LinkedList");
            System.out.println("Press 7 : Display/Traversal the LinkedList");
            System.out.println("Press 8 : Search in LinkedList");
            System.out.println("Press 9 : Exit");
            System.out.println("Enter your choice : ");
            try {
                //  Create a scanner to take input from the user
                Scanner sc = new Scanner(System.in);
                int  choice = sc.nextInt();
                //  Taking the choice from the user and performing the corresponding operation
                switch (choice) {
                    case 0:
                        System.out.print("Please Enter the first Node Value: ");
                        head = createList(sc.nextInt());
                        break;
                    case 1:
                        if (head != null) {
                            System.out.print("Please Enter the Node Value: ");
                            head = insertAtEnd(head,sc.nextInt());
                        } else {
                            System.out.println("Please create a LinkedList first");
                        }
                            break;
                    case 2:
                            if (head != null) {
                                System.out.print("Please Enter the Node Value: ");
                                head = insertAtFront(head,sc.nextInt());
                            } else {
                                System.out.println("Please create a LinkedList first");
                            }
                            break;
                    case 3:
                        if (head != null) {
                            System.out.print("Please Enter the Node Value and Position respectively: ");
                            head = insertAtPosition(head, sc.nextInt(), sc.nextInt());
                        } else {
                            System.out.println("Please create a LinkedList first");
                        }
                        break;
                    case 4:
                        if(head == null){
                            System.out.println("Please create a LinkedList first");
                        } else {
                            System.out.print("Please Enter the Node Value: ");
                            head = deleteNode(head, sc.nextInt());
                        }
                        break;
                    case 5:
                        if (head == null) {
                            System.out.println("Please create a LinkedList first");
                        } else {
                            System.out.print("Please Enter the Node Value: ");
                            System.out.println("Please Enter the new value of the node: ");
                            updateNode(head, sc.nextInt(), sc.nextInt());
                        }
                        break;
                    case 6:
                        if (head == null) {
                            System.out.println("Please create a LinkedList first");
                        } else {
                            head = sortList(head);
                            System.out.println("The Sorted Linked List is : ");
                            displayList(head);
                        }
                        break;
                    case 7:
                        if (head == null) {
                            System.out.println("Please create a LinkedList first");
                        } else {
                            System.out.println("The Linked List is : ");
                            displayList(head);
                        }
                        break;
                    case 8:
                        if (head == null) {
                            System.out.println("Please create a LinkedList first");
                        } else {
                            System.out.print("Please Enter the Node Value: ");
                            searchNode(head, sc.nextInt());
                        }
                        break;
                    case 9:
                        System.out.println("Program Terminated Successfully");
                        System.exit(0);
                        break;
                    default:
                        System.out.println("Invalid Input");
                }
            }
            //  Catching the exception if the user enters a non integer value
            catch(InputMismatchException e){
                System.out.println("Please Enter Integer Choice Only");
            }
        }
    }
    //  Method to create a new linked list
    static Node createList(int value){
        return new Node(value);
    }
    //    Method to add a new node to the linked list
    static Node insertAtEnd(Node head, int data){
        Node temp  = new Node(data);
        Node node = head;
        while(node.next!=null)
        {
            node = node.next;
        }
        node.next = temp;
        return head;
    }
    // Method to insert at front
    static Node insertAtFront(Node head, int data){
        Node temp = new Node(data);
        temp.next = head;
        return temp;
    }
    // Method to insert at position
    static Node insertAtPosition(Node head, int data, int position){
        Node temp = new Node(data);
        Node node = head;
        if(position == 0) {
            temp.next = head;
            return temp;
        }
        for(int i=0;i<position-1;i++){
            node = node.next;
        }
        temp.next = node.next;
        node.next = temp;
        return head;
    }
    //  Method to delete a node from the linked list
    static void displayList(Node head){
        Node temp =  head;
        while(temp != null){
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }
    //  Method to update a node in the linked list
    static Node deleteNode(Node head,int data){
        Node temp = head;
        while(temp.next != null){
            if(temp.next.data == data){
                temp.next = temp.next.next;
                break;
            }
            temp = temp.next;
        }
        displayList(head);
        return head;
    }
    // Method of update note in linked list
    static void updateNode(Node head,int data,int newValue){
        Node temp = head;
        while(temp.next != null){
            if(temp.next.data == data){
                temp.next.data = newValue;
                break;
            }
            temp = temp.next;
        }
        displayList(head);
    }
    // Merge sort on Linked List
    static Node sortList(Node head){
        if(head==null || head.next==null){
            return head;
        }
        Node mid = getMid(head);
        Node nextToMid = mid.next;
        mid.next = null;
        Node left = sortList(head);
        Node right = sortList(nextToMid);
        return merge(left,right);
    }
    //  Function to get the middle node of the linked list : Used in merge sort
    static Node getMid(Node head){
        if(head==null){
            return head;
        }
        Node slow = head;
        Node fast = head.next;
        while(fast != null){
            fast = fast.next;
            if(fast != null){
                slow = slow.next;
                fast = fast.next;
            }
        }
        return slow;
    }
    //    Function to merge two sorted linked lists : Used in merge sort
    static Node merge(Node left, Node right){
        Node result = null;
        if(left == null){
            return right;
        }
        if(right == null){
            return left;
        }
        if(left.data <= right.data){
            result = left;
            result.next = merge(left.next, right);
        }
        else{
            result = right;
            result.next = merge(left, right.next);
        }
        return result;
    }
    //    Function to search a node in the linked list : Linear Search
    static void searchNode(Node head,int data){
        Node temp = head;
        while(temp!=null){
            if(temp.data==data){
                System.out.println("Node Found");
                return;
            }
            temp = temp.next;
        }
        System.out.println("Node Not Found");
    }
}


